File: polyglot/go/http_endpoint_analyzer.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"strings"
	"time"
)

// C2Signature holds the known signatures for each C2 framework
type C2Signature struct {
	Name        string
	Paths       []string
	Headers     map[string]string
	Patterns    []string
	VersionKeys []string
	Timeout     time.Duration
}

var c2Signatures = []C2Signature{
	// Cobalt Strike
	{
		Name:   "Cobalt Strike",
		Paths:  []string{"beacon", "beacon/", "c2/", "/beacon"},
		Headers: map[string]string{
			"X-Beacon-Version": "4.0",
		},
		Patterns: []string{
			`"version":"4\.0`,
			`"type":"beacon",`,
		},
		VersionKeys: []string{"X-Beacon-Version"},
		Timeout:     2 * time.Second,
	},

	// Sliver
	{
		Name:   "Sliver",
		Paths:  []string{"api/v1/auth/login", "api/v1/session", "/api/v1/auth/login"},
		Headers: map[string]string{
			"X-Sliver-Version": "0.12.0",
		},
		Patterns: []string{
			`"version":"0\.12`,
			`"type":"sliver",`,
		},
		VersionKeys: []string{"X-Sliver-Version"},
		Timeout:     3 * time.Second,
	},

	// Mythic
	{
		Name:   "Mythic",
		Paths:  []string{"mythic/", "api/mythic/", "/mythic/"},
		Headers: map[string]string{
			"X-Mythic-Version": "2.0.0",
		},
		Patterns: []string{
			`"version":"2\.0`,
			`"type":"mythic",`,
		},
		VersionKeys: []string{"X-Mythic-Version"},
		Timeout:     3 * time.Second,
	},

	// Havoc
	{
		Name:   "Havoc",
		Paths:  []string{"api/v1/auth/login", "api/v1/session", "/api/v1/auth/login"},
		Headers: map[string]string{
			"X-Havoc-Version": "0.8.0",
		},
		Patterns: []string{
			`"version":"0\.8`,
			`"type":"havoc",`,
		},
		VersionKeys: []string{"X-Havoc-Version"},
		Timeout:     3 * time.Second,
	},

	// Brute Ratel
	{
		Name:   "Brute Ratel",
		Paths:  []string{"ratel/", "api/ratel/", "/ratel/"},
		Headers: map[string]string{
			"X-Ratel-Version": "1.0.0",
		},
		Patterns: []string{
			`"version":"1\.0`,
			`"type":"ratel",`,
		},
		VersionKeys: []string{"X-Ratel-Version"},
		Timeout:     2 * time.Second,
	},

	// Generic fallback - common C2 patterns
	{
		Name:   "Generic C2",
		Paths:  []string{"/api/v1/", "/api/management/"},
		Patterns: []string{
			`"type":"beacon|sliver|mythic|havoc|ratel",`,
			`"protocol":"http|https",`,
		},
		VersionKeys: []string{},
		Timeout:     5 * time.Second,
	},
}

// DetectionResult holds the result of analyzing a single C2 signature
type DetectionResult struct {
	Name        string
	Confidence   float64
	MatchType    string // path, header, pattern, or generic
	Matches      []string
	VersionFound string
	TimeoutHit   bool
}

// AnalysisResult holds the complete analysis of a target URL
type AnalysisResult struct {
	URL         string
	StartTime   time.Time
	EndTime     time.Time
	Duration    time.Duration
	Signatures  []DetectionResult
	TopMatch    *DetectionResult
	HTTPInfo    HTTPInfo
}

// HTTPInfo holds the raw HTTP response information
type HTTPInfo struct {
	Status       int
	Headers      map[string][]string
	Body         string
	ContentType  string
	ContentLength int64
	TimeoutHit   bool
}

// HTTPAnalyzer handles the HTTP analysis for C2 detection
type HTTPAnalyzer struct {
	client *http.Client
}

func NewHTTPAnalyzer() *HTTPAnalyzer {
	return &HTTPAnalyzer{
		client: &http.Client{
			Transport: &http.Transport{
				DialContext: (&net.Dialer{
					Timeout:   30 * time.Second,
					DualStack: true,
				}).DialContext,
				MaxIdleConns:        100,
				MaxIdleConnsPerHost: 10,
				IdleConnTimeout:     90 * time.Second,
			},
			Timeout:   30 * time.Second,
			CheckRedirect: func(req *http.Request, via []*http.Request) error {
				if len(via) >= 5 {
					return fmt.Errorf("too many redirects")
				}
				return nil
			},
		},
	}
}

func (a *HTTPAnalyzer) Analyze(targetURL string) (*AnalysisResult, error) {
	startTime := time.Now()
	
	result := &AnalysisResult{
		URL:      targetURL,
		StartTime: startTime,
		HTTPInfo: HTTPInfo{},
	}

	// Parse the URL and try multiple paths
	parsedURL, err := url.Parse(targetURL)
	if err != nil {
		return result, fmt.Errorf("failed to parse URL: %w", err)
	}

	host := parsedURL.Host
	path := parsedURL.Path

	// Try the base path first
	result.HTTPInfo = a.fetchResponse(host, path)
	
	// If no response or empty body, try common C2 paths
	if result.HTTPInfo.Status == 0 || result.HTTPInfo.Body == "" {
		commonPaths := []string{"/beacon", "/api/v1/auth/login", "/mythic/", "/ratel/"}
		
		for _, cp := range commonPaths {
			result.HTTPInfo = a.fetchResponse(host, cp)
			if result.HTTPInfo.Status != 0 && len(result.HTTPInfo.Body) > 0 {
				break
			}
		}
	}

	// Analyze the response against all signatures
	for _, sig := range c2Signatures {
		result.Signatures = append(result.Signatures, a.analyzeSignature(sig, result.HTTPInfo))
	}

	result.EndTime = time.Now()
	result.Duration = result.EndTime.Sub(startTime)

	// Find the top match
	if len(result.Signatures) > 0 {
		sort.Slice(result.Signatures, func(i, j int) bool {
			return result.Signatures[i].Confidence > result.Signatures[j].Confidence
		})
		result.TopMatch = &result.Signatures[0]
	}

	return result, nil
}

func (a *HTTPAnalyzer) fetchResponse(host, path string) HTTPInfo {
	var resp HTTPInfo
	
	reqURL := fmt.Sprintf("http://%s%s", host, path)
	if strings.HasPrefix(path, "/") && !strings.Contains(reqURL, "://") {
		reqURL = reqURL[7:] + path // remove leading / if present
	}

	resp.Status = 0
	resp.TimeoutHit = false
	
	var body []byte
	var err error
	
	// Try HTTP first, then HTTPS
	for _, scheme := range []string{"http", "https"} {
		urlToTry := fmt.Sprintf("%s://%s%s", scheme, host, path)
		
		req, err := http.NewRequest("GET", urlToTry, nil)
		if err != nil {
			continue
		}

		resp, err = a.client.Do(req)
		if err == nil && resp != nil {
			body, err = io.ReadAll(resp.Body)
			if err == nil {
				resp.Status = resp.StatusCode
				resp.Headers = make(map[string][]string)
				for k, v := range resp.Header {
					resp.Headers[k] = append(resp.Headers[k], v...)
				}
				resp.ContentType = strings.Split(resp.Header.Get("Content-Type"), ";")[0]
				resp.ContentLength = int64(len(body))
				resp.Body = string(body)
				resp.TimeoutHit = resp.StatusCode >= 500 || err != nil
				
				if resp.Status > 0 {
					return resp
				}
			}
		}
		
		if resp.Status > 0 {
			break
		}
	}

	return resp
}

func (a *HTTPAnalyzer) analyzeSignature(sig C2Signature, httpInfo HTTPInfo) DetectionResult {
	result := DetectionResult{
		Name:        sig.Name,
		Confidence:   0.0,
		TimeoutHit:   httpInfo.TimeoutHit,
	}

	if httpInfo.Status == 0 || len(httpInfo.Body) == 0 {
		return result
	}

	// Check paths
	for _, p := range sig.Paths {
		if strings.Contains(httpInfo.Body, p) || 
		   strings.Contains(httpInfo.URL, p) {
			result.Confidence = max(result.Confidence, 0.6)
			result.MatchType = "path"
			result.Matches = append(result.Matches, fmt.Sprintf("path match: %s", p))
		}
	}

	// Check headers
	for k, v := range sig.Headers {
		if strings.Contains(httpInfo.Body, k) || 
		   httpInfo.Headers[k] != "" {
			result.Confidence = max(result.Confidence, 0.7)
			result.MatchType = "header"
			result.Matches = append(result.Matches, fmt.Sprintf("header match: %s", k))
		}
	}

	// Check patterns in body
	for _, pattern := range sig.Patterns {
		if strings.Contains(httpInfo.Body, pattern) {
			result.Confidence = max(result.Confidence, 0.8)
			result.MatchType = "pattern"
			result.Matches = append(result.Matches, fmt.Sprintf("pattern match: %s", pattern))
		}
	}

	// Check version keys in headers
	for _, key := range sig.VersionKeys {
		if httpInfo.Headers[key] != "" {
			result.Confidence = max(result.Confidence, 0.85)
			result.MatchType = "version"
			result.Matches = append(result.Matches, fmt.Sprintf("version header: %s=%s", key, httpInfo.Headers[key]))
		}
	}

	return result
}

func (a *HTTPAnalyzer) GetTopMatch(result *AnalysisResult) string {
	if result.TopMatch == nil || result.TopMatch.Confidence == 0 {
		return "Unknown"
	}
	
	name := result.TopMatch.Name
	
	// Add version if found
	if result.TopMatch.VersionFound != "" {
		name += fmt.Sprintf(" v%s", result.TopMatch.VersionFound)
	}

	return name
}

func (a *HTTPAnalyzer) PrintResult(result *AnalysisResult) {
	fmt.Printf("\n=== C2 Detection Analysis ===\n")
	fmt.Printf("Target: %s\n", result.URL)
	fmt.Printf("Duration: %.3fs\n", result.Duration.Seconds())
	fmt.Printf("Status: %d\n", result.HTTPInfo.Status)
	
	if len(result.Signatures) > 0 {
		fmt.Printf("\n--- Detected Signatures (sorted by confidence) ---\n")
		
		for i, sig := range result.Signatures {
			confidenceStr := fmt.Sprintf("%.1f%%", sig.Confidence*100)
			if sig.TimeoutHit {
				confidenceStr += " [timeout]"
			}
			
			fmt.Printf("\n%d. %s\n", i+1, sig.Name)
			fmt.Printf("   Confidence: %s\n", confidenceStr)
			fmt.Printf("   Type: %s\n", sig.MatchType)
			
			if len(sig.Matches) > 0 {
				for _, m := range sig.Matches {
					fmt.Printf("   - %s\n", m)
				}
			}
		}
		
		if result.TopMatch != nil && result.TopMatch.Confidence > 0.5 {
			fmt.Printf("\n--- TOP MATCH ---\n")
			fmt.Printf("%s (Confidence: %.1f%%)\n", 
				result.GetTopMatch(result), 
				result.TopMatch.Confidence*100)
		}
	} else {
		fmt.Println("No signatures detected.")
		
		if result.HTTPInfo.Status == 200 && len(result.HTTPInfo.Body) > 0 {
			fmt.Printf("\n--- Raw Response (first 500 chars) ---\n")
			fmt.Println(result.HTTPInfo.Body[:min(500, len(result.HTTPInfo.Body))])
		} else if result.HTTPInfo.Status != 0 {
			fmt.Printf("Status: %d\n", result.HTTPInfo.Status)
		}
	}
	
	if result.HTTPInfo.ContentType != "" {
		fmt.Printf("\n--- Content-Type ---\n")
		fmt.Println(result.HTTPInfo.ContentType)
	}
}

func main() {
	analyzer := NewHTTPAnalyzer()
	
	// Default target - can be overridden via command line
	targetURL := "http://localhost:8080/beacon"
	if len(os.Args) > 1 {
		targetURL = os.Args[1]
	}

	fmt.Printf("C2 Endpoint Analyzer v1.0\n")
	fmt.Printf("Analyzing: %s\n", targetURL)
	
	result, err := analyzer.Analyze(targetURL)
	if err != nil {
		fmt.Printf("Error during analysis: %v\n", err)
		os.Exit(1)
	}

	analyzer.PrintResult(result)
	
	// Exit with appropriate code based on detection
	if result.TopMatch != nil && result.TopMatch.Confidence > 0.5 {
		fmt.Printf("\nFinal verdict: %s detected!\n", analyzer.GetTopMatch(result))
		os.Exit(0)
	} else if len(result.Signatures) > 0 {
		fmt.Println("Partial matches found - verify manually.")
		os.Exit(1)
	} else {
		fmt.Println("No clear C2 signature detected.")
		os.Exit(2)
	}
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func max(a, b float64) float64 {
	if a > b {
		return a
	}
	return b
}

File: polyglot/python/http_endpoint_analyzer.py
import socket
import ssl
import re
from typing import Dict, Any, Optional, Tuple
from urllib.parse import urlparse
from datetime import datetime


class C2Fingerprint:
    """Core fingerprinting engine for C2 servers."""
    
    # Known header patterns and their likelihood scores
    HEADER_SIGNATURES = {
        'cobalt_strike': [
            r'X-Cobalt-Strike',
            r'Server: Cobalt',
            r'X-Forwarded-For:\s*\d+\.\d+\.\d+\.\d+',  # Often has multiple hops
        ],
        'sliver': [
            r'Sliver',
            r'x-sliver',
            r'Server: Sliver',
        ],
        'mythic': [
            r'Mythic',
            r'X-Mythic',
            r'mythic\.io',
        ],
        'havoc': [
            r'Havoc',
            r'x-havoc',
            r'havoc\.net',
        ],
        'brute_ratel': [
            r'BruteRatel',
            r'x-bruteratel',
            r'ratel\.io',
        ],
    }

    # Body content signatures (partial responses)
    BODY_SIGNATURES = {
        'cobalt_strike': [
            b'cobaltstrike',
            b'Cobalt Strike Beacon',
            b'storm/beacon',
        ],
        'sliver': [
            b'sliver',
            b'Sliver C2',
            b'sliver\.io',
        ],
        'mythic': [
            b'mythic',
            b'Mythic C2',
            b'mythic\.io/api',
        ],
        'havoc': [
            b'havoc',
            b'Havoc C2',
            b'havoc\.net',
        ],
        'brute_ratel': [
            b'BruteRatel',
            b'ratel\.io',
        ],
    }

    # TLS fingerprint hints (simplified)
    TLS_HINTS = {
        'cobalt_strike': {'cipher_suites': ['TLS_AES_256_GCM_SHA384'], 'version': '1.2'},
        'sliver': {'cipher_suites': ['ECDHE-RSA-AES256-GCM-SHA384'], 'version': '1.2'},
    }


class HTTPEndpointAnalyzer:
    """Complete HTTP endpoint analyzer for C2 server fingerprinting."""

    def __init__(self, timeout: float = 5.0):
        self.timeout = timeout
        self.results_cache: Dict[str, Any] = {}
    
    def _create_connection(
        self, 
        host: str, 
        port: int, 
        use_ssl: bool = False,
        headers: Optional[Dict[str, str]] = None
    ) -> Tuple[socket.socket, ssl.SSLContext | None]:
        """Establish connection with proper SSL handling."""
        if use_ssl and port in (443, 8443):
            context = ssl.create_default_context()
            try:
                context.load_verify_locations('/etc/ssl/certs/ca-certificates.crt')
            except FileNotFoundError:
                pass
            
            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            sock.settimeout(self.timeout)
            sock.connect((host, port))
            
            # Wrap with SSL
            ssl_sock = context.wrap_socket(sock, server_hostname=host)
            return ssl_sock, context
        
        else:
            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            sock.settimeout(self.timeout)
            sock.connect((host, port))
            return sock, None
    
    def _extract_headers(self, raw_response: bytes) -> Dict[str, str]:
        """Parse HTTP headers from raw response."""
        try:
            text = raw_response.decode('utf-8', errors='ignore')
            lines = text.split('\r\n\r\n')[0].split('\r\n')
            
            headers = {}
            for line in lines:
                if ':' in line:
                    key, value = line.split(':', 1)
                    headers[key.strip().lower()] = value.strip()
            
            return headers
        except Exception:
            return {}

    def _extract_body_sample(self, raw_response: bytes, max_size: int = 4096) -> str:
        """Get a safe sample of response body for pattern matching."""
        try:
            text = raw_response.decode('utf-8', errors='ignore')[:max_size]
            return text.lower()
        except Exception:
            return ''

    def _match_header_signatures(
        self, 
        headers: Dict[str, str],
        sample_body: str
    ) -> Tuple[bool, list]:
        """Check headers against known C2 signatures."""
        matches = []
        
        for c2_type, patterns in C2Fingerprint.HEADER_SIGNATURES.items():
            for pattern in patterns:
                if re.search(pattern, sample_body, re.IGNORECASE):
                    matches.append(c2_type)
                    break
        
        return len(matches) > 0, sorted(set(matches))

    def _match_body_signatures(self, sample_body: str) -> Tuple[bool, list]:
        """Check response body for C2 signatures."""
        matches = []
        
        for c2_type, patterns in C2Fingerprint.BODY_SIGNATURES.items():
            for pattern in patterns:
                if re.search(pattern, sample_body, re.IGNORECASE):
                    matches.append(c2_type)
                    break
        
        return len(matches) > 0, sorted(set(matches))

    def _analyze_tls(self, ssl_context: ssl.SSLContext | None) -> Dict[str, Any]:
        """Extract TLS information if available."""
        tls_info = {}
        
        if ssl_context and hasattr(ssl_context, 'version'):
            try:
                version = ssl_context.version()
                tls_info['version'] = version
            except Exception:
                pass
        
        return tls_info

    def _calculate_response_time(self, start_time: float) -> float:
        """Calculate total response time."""
        # This would need to track the end time in a real implementation
        # For now, returns 0 - caller should track this
        return 0.0

    def analyze_endpoint(
        self, 
        url: str, 
        method: str = 'GET',
        headers: Optional[Dict[str, str]] = None
    ) -> Dict[str, Any]:
        """
        Analyze a single HTTP endpoint for C2 fingerprints.
        
        Args:
            url: The target URL (e.g., http://example.com/api)
            method: HTTP method to use (default GET)
            headers: Optional custom headers
            
        Returns:
            Dictionary containing analysis results
        """
        parsed = urlparse(url)
        host = parsed.hostname or 'unknown'
        port = parsed.port or 80 if not parsed.scheme.startswith('https') else 443
        
        # Determine if SSL is needed
        use_ssl = parsed.scheme == 'https' and (port in (443, 8443) or 
                                                ('443' in str(parsed.netloc)))

        try:
            sock, ssl_context = self._create_connection(host, port, use_ssl, headers)
            
            # Build request
            path = parsed.path + (parsed.query if parsed.query else '')
            body = f"{method} {path} HTTP/1.1\r\n"
            body += "Host: " + host + "\r\n"
            body += "Connection: close\r\n"
            
            if headers:
                for k, v in headers.items():
                    body += f"{k}: {v}\r\n"
            
            # Add User-Agent to help detect server behavior
            user_agent = headers.get('User-Agent', '') if headers else ''
            if not user_agent:
                body += "User-Agent: Python-C2-Detect/1.0\r\n"
            
            body += "\r\n"
            
            sock.sendall(body.encode())
            
            # Read response with size limit
            sock.settimeout(5.0)
            buffer = b''
            while True:
                chunk = sock.recv(4096)
                if not chunk:
                    break
                buffer += chunk
                if b'\r\n\r\n' in buffer and len(buffer) > 2048:
                    # Found headers, get body sample
                    header_end = buffer.index(b'\r\n\r\n') + 4
                    body_sample = buffer[header_end:].decode('utf-8', errors='ignore')[:1024]
                    break
            
            sock.close()

        except socket.timeout:
            return {
                'status': 'timeout',
                'host': host,
                'port': port,
                'error': f'Connection timed out after {self.timeout}s',
                'headers': {},
                'body_sample': '',
                'tls_info': {},
                'response_time_ms': 0.0,
            }

        except socket.error as e:
            return {
                'status': 'error',
                'host': host,
                'port': port,
                'error': str(e),
                'headers': {},
                'body_sample': '',
                'tls_info': {},
                'response_time_ms': 0.0,
            }

        except Exception as e:
            return {
                'status': 'exception',
                'host': host,
                'port': port,
                'error': type(e).__name__ + ': ' + str(e),
                'headers': {},
                'body_sample': '',
                'tls_info': {},
                'response_time_ms': 0.0,
            }

        # Process successful response
        headers = self._extract_headers(buffer)
        body_sample = self._extract_body_sample(buffer)

        # Run all analysis modules
        header_match, header_matches = self._match_header_signatures(headers, body_sample)
        body_match, body_matches = self._match_body_signatures(body_sample)
        tls_info = self._analyze_tls(ssl_context)

        # Calculate confidence score (0-100)
        confidence = 0.0
        if header_match:
            confidence += 30
        if body_match:
            confidence += 40
        if tls_info.get('version'):
            confidence += 10
        
        # Adjust for number of matches
        total_matches = len(header_matches) + len(body_matches)
        confidence = min(100, confidence + (total_matches * 5))

        return {
            'status': 'success',
            'host': host,
            'port': port,
            'url': url,
            'headers': headers,
            'body_sample': body_sample,
            'tls_info': tls_info,
            'response_time_ms': 0.0,  # Would need tracking in real impl
            'header_matches': header_match,
            'header_candidates': header_matches,
            'body_matches': body_match,
            'body_candidates': body_matches,
            'confidence_score': round(confidence, 1),
            'is_c2_server': header_match or body_match,
        }


class C2ServerScanner:
    """High-level scanner that orchestrates multiple endpoint analyses."""

    def __init__(self, timeout: float = 5.0):
        self.analyzer = HTTPEndpointAnalyzer(timeout=timeout)
    
    def scan_endpoint(self, url: str, method: str = 'GET') -> Dict[str, Any]:
        """Scan a single URL and return detailed results."""
        result = self.analyzer.analyze_endpoint(url, method)
        
        # Add summary fields
        result['summary'] = {
            'is_c2_server': result.get('is_c2_server', False),
            'likely_type': max(result.get('header_candidates', []), 
                             key=lambda x: 1 if x in result.get('header_candidates', []) else 0)
                            or max(result.get('body_candidates', []))
                            or 'unknown'
                            if (result.get('header_matches') or result.get('body_matches'))
                            else 'unknown',
            'confidence': result.get('confidence_score', 0.0),
        }

        return result
    
    def scan_multiple(self, urls: list[str], method: str = 'GET') -> list[Dict]:
        """Scan multiple URLs and aggregate results."""
        results = []
        
        for url in urls:
            try:
                result = self.scan_endpoint(url, method)
                results.append(result)
                
                # Print quick summary
                if result.get('is_c2_server'):
                    print(f"  [!] C2 detected at {url}")
                    print(f"      Type: {result['summary']['likely_type']}")
                    print(f"      Confidence: {result['summary']['confidence']}%")
                    
            except Exception as e:
                results.append({
                    'status': 'exception',
                    'host': urlparse(url).hostname or url,
                    'error': str(e),
                })

        return results


def main():
    """Demo/test harness for the analyzer."""
    
    print("=" * 60)
    print("C2 Server Fingerprinter - HTTP Endpoint Analyzer")
    print("=" * 60)
    print()
    
    # Test URLs (some real patterns, some examples)
    test_urls = [
        'http://example.com/api',           # Generic API endpoint
        'https://api.github.com/user',      # Real HTTPS with cert
        'http://127.0.0.1:8080/test',       # Local test
    ]

    print("Testing HTTPEndpointAnalyzer...")
    print("-" * 40)
    
    analyzer = HTTPEndpointAnalyzer(timeout=3.0)
    
    for url in test_urls[:2]:  # Limit for demo speed
        print(f"\nAnalyzing: {url}")
        result = analyzer.analyze_endpoint(url, method='GET')
        
        if result.get('status') == 'success':
            print(f"  Status: OK")
            print(f"  Host: {result['host']}:{result['port']}")
            print(f"  Is C2 Server: {result.get('is_c2_server', False)}")
            print(f"  Confidence: {result.get('confidence_score', 0):.1f}%")
            
            if result.get('header_candidates'):
                print(f"  Header Matches: {', '.join(result['header_candidates'])}")
            if result.get('body_candidates'):
                print(f"  Body Matches: {', '.join(result['body_candidates'])}")

    # Test with scanner
    print("\n\nTesting C2ServerScanner...")
    print("-" * 40)
    
    scanner = C2ServerScanner(timeout=3.0)
    results = scanner.scan_multiple(test_urls, method='GET')
    
    print(f"\nTotal scanned: {len(results)}")
    c2_count = sum(1 for r in results if r.get('is_c2_server'))
    print(f"C2 servers detected: {c2_count}")

    # Interactive mode suggestion
    print("\n" + "=" * 60)
    print("Ready to analyze endpoints.")
    print("Use C2ServerScanner().scan_multiple([url1, url2, ...])")
    print("=" * 60)


if __name__ == '__main__':
    main()